File: aether_messenger/android/app/src/main/java/com/aetherq/messenger/mqtt/MessageEnvelope.kt
package com.aetherq.messenger.mqtt

import uniffi.aether_ffi.AkemEncapsulation
import java.io.ByteArrayOutputStream

private const val TYPE_HANDSHAKE: Byte = 0x01
private const val TYPE_MESSAGE: Byte = 0x02

private const val USER_ID_LEN = 8
private const val CT_MLKEM_LEN = 1088 // ML-KEM-768 ciphertext — spec 2-bo'lim jadvali, qat'iy o'lcham
private const val CT_X25519_LEN = 32
private const val SIGNATURE_LEN = 64 // Ed25519 — qat'iy o'lcham
private const val COMPRESS_FRAME_LEN = 77 // spec 4.2-bo'lim

sealed class InboundEnvelope {
    data class Handshake(
        val senderUserId: ByteArray,
        val ctMlkem: ByteArray,
        val ctX25519: ByteArray,
        val signature: ByteArray,
    ) : InboundEnvelope()

    data class RatchetMessage(
        val compressFrame: ByteArray,
        val recordFrame: ByteArray,
    ) : InboundEnvelope()
}

/**
 * MQTT payload'ga qadab yuboriladigan, ilova darajasidagi binary frame format.
 * Har bir maydon spec-primitivlaridan kelib chiqqan qat'iy o'lchamga ega bo'lgani
 * uchun (ML-KEM-768 ct, Ed25519 sig, Compress-KEM frame — barchasi belgilangan
 * o'lchamda) uzunlik-prefikslar kerak emas — faqat oxirgi maydon (record ciphertext)
 * o'zgaruvchan uzunlikka ega va "qolgan barcha baytlar" sifatida o'qiladi.
 */
object MessageEnvelope {

    fun encodeHandshake(senderUserId: ByteArray, encaps: AkemEncapsulation): ByteArray {
        require(senderUserId.size == USER_ID_LEN) { "senderUserId $USER_ID_LEN bayt bo'lishi SHART" }
        require(encaps.ctMlkem.size == CT_MLKEM_LEN) { "ct_mlkem kutilmagan uzunlik: ${encaps.ctMlkem.size}" }
        require(encaps.ctX25519.size == CT_X25519_LEN) { "ct_x25519 kutilmagan uzunlik: ${encaps.ctX25519.size}" }
        require(encaps.signature.size == SIGNATURE_LEN) { "signature kutilmagan uzunlik: ${encaps.signature.size}" }

        val out = ByteArrayOutputStream(1 + USER_ID_LEN + CT_MLKEM_LEN + CT_X25519_LEN + SIGNATURE_LEN)
        out.write(TYPE_HANDSHAKE.toInt())
        out.write(senderUserId)
        out.write(encaps.ctMlkem)
        out.write(encaps.ctX25519)
        out.write(encaps.signature)
        return out.toByteArray()
    }

    fun encodeMessage(compressFrame: ByteArray, recordFrame: ByteArray): ByteArray {
        require(compressFrame.size == COMPRESS_FRAME_LEN) { "compress_frame 77 bayt bo'lishi SHART" }
        val out = ByteArrayOutputStream(1 + COMPRESS_FRAME_LEN + recordFrame.size)
        out.write(TYPE_MESSAGE.toInt())
        out.write(compressFrame)
        out.write(recordFrame)
        return out.toByteArray()
    }

    fun decode(payload: ByteArray): InboundEnvelope? {
        if (payload.isEmpty()) return null
        return when (payload[0]) {
            TYPE_HANDSHAKE -> decodeHandshake(payload)
            TYPE_MESSAGE -> decodeMessage(payload)
            else -> null
        }
    }

    private fun decodeHandshake(payload: ByteArray): InboundEnvelope.Handshake? {
        val expected = 1 + USER_ID_LEN + CT_MLKEM_LEN + CT_X25519_LEN + SIGNATURE_LEN
        if (payload.size != expected) return null

        var offset = 1
        val userId = payload.copyOfRange(offset, offset + USER_ID_LEN); offset += USER_ID_LEN
        val ctMlkem = payload.copyOfRange(offset, offset + CT_MLKEM_LEN); offset += CT_MLKEM_LEN
        val ctX25519 = payload.copyOfRange(offset, offset + CT_X25519_LEN); offset += CT_X25519_LEN
        val signature = payload.copyOfRange(offset, offset + SIGNATURE_LEN)
        return InboundEnvelope.Handshake(userId, ctMlkem, ctX25519, signature)
    }

    private fun decodeMessage(payload: ByteArray): InboundEnvelope.RatchetMessage? {
        val minLen = 1 + COMPRESS_FRAME_LEN
        if (payload.size < minLen) return null
        val compressFrame = payload.copyOfRange(1, 1 + COMPRESS_FRAME_LEN)
        val recordFrame = payload.copyOfRange(1 + COMPRESS_FRAME_LEN, payload.size)
        return InboundEnvelope.RatchetMessage(compressFrame, recordFrame)
    }
}

File: aether_messenger/windows/AetherQMessenger/Mqtt/MessageEnvelope.cs
using System;
using System.IO;
using AetherQMessenger.Ffi;

namespace AetherQMessenger.Mqtt;

public abstract record InboundEnvelope
{
    public sealed record Handshake(byte[] SenderUserId, byte[] CtMlkem, byte[] CtX25519, byte[] Signature) : InboundEnvelope;

    public sealed record RatchetMessage(byte[] CompressFrame, byte[] RecordFrame) : InboundEnvelope;
}

/// <summary>
/// MQTT payload'ga qadab yuboriladigan binary frame format — Android (Kotlin)
/// versiyasidagi <c>MessageEnvelope.kt</c> bilan bayt-baytiga bir xil, shu tufayli
/// ikkala klient (mobil/desktop) o'rtasida haqiqiy interop mumkin.
///
/// Har bir maydon spec-primitivlaridan kelib chiqqan qat'iy o'lchamga ega (ML-KEM-768 ct,
/// Ed25519 sig, Compress-KEM frame) — uzunlik-prefikslar kerak emas.
/// </summary>
public static class MessageEnvelope
{
    private const byte TypeHandshake = 0x01;
    private const byte TypeMessage = 0x02;

    private const int UserIdLen = 8;
    private const int CtMlkemLen = 1088;
    private const int CtX25519Len = 32;
    private const int SignatureLen = 64;
    private const int CompressFrameLen = 77;

    public static byte[] EncodeHandshake(byte[] senderUserId, AkemEncapsulation encaps)
    {
        if (senderUserId.Length != UserIdLen)
        {
            throw new ArgumentException($"senderUserId {UserIdLen} bayt bo'lishi SHART", nameof(senderUserId));
        }
        if (encaps.CtMlkem.Length != CtMlkemLen)
        {
            throw new ArgumentException($"ct_mlkem kutilmagan uzunlik: {encaps.CtMlkem.Length}");
        }
        if (encaps.CtX25519.Length != CtX25519Len)
        {
            throw new ArgumentException($"ct_x25519 kutilmagan uzunlik: {encaps.CtX25519.Length}");
        }
        if (encaps.Signature.Length != SignatureLen)
        {
            throw new ArgumentException($"signature kutilmagan uzunlik: {encaps.Signature.Length}");
        }

        using var stream = new MemoryStream(1 + UserIdLen + CtMlkemLen + CtX25519Len + SignatureLen);
        stream.WriteByte(TypeHandshake);
        stream.Write(senderUserId);
        stream.Write(encaps.CtMlkem);
        stream.Write(encaps.CtX25519);
        stream.Write(encaps.Signature);
        return stream.ToArray();
    }

    public static byte[] EncodeMessage(byte[] compressFrame, byte[] recordFrame)
    {
        if (compressFrame.Length != CompressFrameLen)
        {
            throw new ArgumentException("compress_frame 77 bayt bo'lishi SHART", nameof(compressFrame));
        }

        using var stream = new MemoryStream(1 + CompressFrameLen + recordFrame.Length);
        stream.WriteByte(TypeMessage);
        stream.Write(compressFrame);
        stream.Write(recordFrame);
        return stream.ToArray();
    }

    public static InboundEnvelope? Decode(byte[] payload)
    {
        if (payload.Length == 0)
        {
            return null;
        }

        return payload[0] switch
        {
            TypeHandshake => DecodeHandshake(payload),
            TypeMessage => DecodeMessage(payload),
            _ => null,
        };
    }

    private static InboundEnvelope.Handshake? DecodeHandshake(byte[] payload)
    {
        const int expected = 1 + UserIdLen + CtMlkemLen + CtX25519Len + SignatureLen;
        if (payload.Length != expected)
        {
            return null;
        }

        var offset = 1;
        var userId = payload[offset..(offset + UserIdLen)]; offset += UserIdLen;
        var ctMlkem = payload[offset..(offset + CtMlkemLen)]; offset += CtMlkemLen;
        var ctX25519 = payload[offset..(offset + CtX25519Len)]; offset += CtX25519Len;
        var signature = payload[offset..(offset + SignatureLen)];
        return new InboundEnvelope.Handshake(userId, ctMlkem, ctX25519, signature);
    }

    private static InboundEnvelope.RatchetMessage? DecodeMessage(byte[] payload)
    {
        const int minLen = 1 + CompressFrameLen;
        if (payload.Length < minLen)
        {
            return null;
        }

        var compressFrame = payload[1..(1 + CompressFrameLen)];
        var recordFrame = payload[(1 + CompressFrameLen)..];
        return new InboundEnvelope.RatchetMessage(compressFrame, recordFrame);
    }
}
